File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class Node:

    def __init__(self, value):
        self.value = value
        self.next = [None] * 26
        self.isEndOfWord = False

class WordDictionary:

    def __init__(self): 
        self.node = Node("")
        

    def addWord(self, word: str) -> None:
        current = self.node
        for c in word:
            if current.next[ord(c) - ord('a')] is None:
                current.next[ord(c) - ord('a')] = Node(c)
            current = current.next[ord(c) - ord('a')]
        current.isEndOfWord = True

    def search(self, word: str) -> bool:
        return self.searchWord(word, 0, self.node)

    
    def searchWord(self, word, index, node):

        if node is None:
            return False

        if index == len(word):
            return node.isEndOfWord

        if word[index] == ".":
            for nextNode in node.next:
                if self.searchWord(word, index+1, nextNode):
                    return True
            return False

        if node.next[ord(word[index]) - ord('a')] is None:
            return False

        return self.searchWord(word, index+1, node.next[ord(word[index]) - ord('a')])

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class Node {

    public char c;
    public Node[] next;
    public boolean isEndOfWord;

    public Node(){
        this.next = new Node[26];
        this.isEndOfWord = false;
    }    

    public Node(char c){
        this.c = c;
        this.next = new Node[26];
        this.isEndOfWord = false;
    }    
}
class WordDictionary {

    private Node root;

    public WordDictionary() {
        this.root = new Node();        
    }
    
    public void addWord(String word) {

        Node current = this.root;

        for(char c : word.toCharArray()){
            if(current.next[c - 'a'] == null){
                current.next[c - 'a'] = new Node(c);
            }
            current = current.next[c - 'a'];
        }
        current.isEndOfWord = true;
    }
    
    public boolean search(String word) {
        return dfs(word, 0, root);
    }

    public boolean dfs(String word, int index, Node current) {

        if(index == word.length()){
            return current.isEndOfWord;
        }

        char c = word.charAt(index);
        
        if(c == '.'){
            for(Node nextNode : current.next){
                if(nextNode != null && dfs(word, index+1, nextNode)){
                    return true;
                }
            }
            return false;
        } else {
            if(current.next[c - 'a'] == null){
                return false;
            }
            current = current.next[c - 'a'];
         }

        return dfs(word, index+1, current); 
    }    
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */